### PR TITLE
Plotting function coloring unstable and stable manifolds using gradient-extraction.  

### PR DESCRIPTION
**Pull request by vkrajnak**
Fixed plotting fuction to show unstable manifolds in red, stable in blue. As a consequence, the gradient of the total LD for 1D centre shows only numerical artefacts - as one would expect in a system with no invariant manifolds.

### CODE REVIEW

**Comment by vkrajnak** [comment]
Any suggestions?

**Comment by broncio123** [comment]
> Any suggestions?

```python
draw_all_lds([], LD_backward, grid_parameters, tau, p_value)
```
Returns

image

```python
draw_all_lds(LD_forward, [], grid_parameters, tau, p_value)
```
image

**Comment by vkrajnak** [comment]
Better?

File: pylds/tools.py [hunk 1]
```diff
@@ -9,40 +9,38 @@
...
+
+    string_title : string
+        Plot title.
+
+    colormap_gradient : string
+        Name of matplotlib colormap for gradient contouor plot.
+
     Returns
     -------
         Nothing.
```
Comment (broncio123) [inline pylds/tools.py:41]: Not sure `Nothing` is appropriate.
There are no guidelines for graphical outputs in here
https://numpydoc.readthedocs.io/en/latest/format.html 

But, in the documentation of [`matplotlib.pyplot.plot`](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.pyplot.plot.html) or [`matplotlib.axes.Axes.contourf`](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.axes.Axes.contourf.html?highlight=contourf#matplotlib.axes.Axes.contourf) they provide a brief description of the graphical output and/or point out to some `matplotlib` object.

File: pylds/tools.py [hunk 2]
```diff
@@ -56,75 +54,109 @@ def draw_lagrangian_descriptor(LD, LD_type, grid_parameters, tau, p_value, color
...
+        str_method = r'LD$_p$ $(p={})$'.format(p_value)
+    t_final=abs(tau)
+
+    # Prepare LD arrays
+    def norm_and_grad(LD):
+        gradient_x, gradient_y = np.gradient(LD)
+        gradient_magnitude = np.sqrt(gradient_x**2 + gradient_y**2)
+        gradient_magnitude = gradient_magnitude - np.nanmin(gradient_magnitude)
+        gradient_magnitude = gradient_magnitude / np.nanmax(gradient_magnitude)
+        return gradient_magnitude
```
[the remark this answers is not in the document]
Comment (broncio123) [reply pylds/tools.py:146]: Nested function. 
Function can be defined outside `draw_all_lds` so that it can be accessed separately.

**Comment by broncio123** [comment]
Nice.
image

**Comment by broncio123** [comment]
Suggestion: In `LD_Python_Examples.ipynb`, add an example using `ld_plot` alone to plot a single LD type with(out) interactivity, assuming someone would like to plot one LD type alone.

**Review by broncio123 — CHANGES_REQUESTED** [review]
Overall the logic behind the new plotting functions is clear and functions read well. 
Use of sliders can lead to neat graphical output as shown in some examples. However, it's clear that some artefacts in the plotting of the LD gradient can occur, e.g., Hamilton centre. 

Address minor fixes and suggestions at your discretion.

File: pylds/tools.py [hunk 3]
```diff
@@ -56,75 +54,109 @@ def draw_lagrangian_descriptor(LD, LD_type, grid_parameters, tau, p_value, color
...
-        @widgets.interact(grad_minimum=(0, .99, .01),grad_maximum=(0.01, 1.0, .01))
-        def update(grad_minimum=0.0,grad_maximum=1.0):
-            grad_max = max(grad_minimum+0.01, grad_maximum)
-            con1.set_clim(grad_minimum,grad_max)
-            
-    ticks_gradient = np.linspace(0,1, 11)
+        @widgets.interact(grad_min=(vmin, vmax-0.01, .01),grad_max=(vmin+0.01, vmax, .01))
+        def update(grad_min=vmin,grad_max=vmax):
+            grad_max = max(grad_min+0.01, grad_max)
+            con1.set_clim(grad_min,grad_max)
```
[the remark this answers is not in the document]
Comment (vkrajnak) [reply pylds/tools.py:84]: Yes, this has to do with `levels=100` in line 79. It is a value that may be used via a hard-coded variable. I don't think there is much point to making it a user-defined variable. What do you think?

File: pylds/tools.py (hunk 1 above)
Comment (vkrajnak) [reply pylds/tools.py:41]: I see what you mean. The matplotlib functions you mention return an object, in contrast to ours. Would `No output` or similar work?

**Comment by vkrajnak** [comment]
> Suggestion: In `LD_Python_Examples.ipynb`, add an example using `ld_plot` alone to plot a single LD type with(out) interactivity, assuming someone would like to plot one LD type alone.

Good idea!

File: pylds/tools.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (vkrajnak) [reply pylds/tools.py]: Done

File: pylds/tools.py (hunk 1 above)
Comment (broncio123) [reply pylds/tools.py:41]: @Shibabrat any thoughts?

File: pylds/tools.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vkrajnak) [reply pylds/tools.py]: done

File: pylds/tools.py [hunk 4]
```diff
@@ -66,7 +66,9 @@ def ld_plot(LD, LD_gradient, grid_parameters, colormap, interactive, string_titl
     plt.subplots_adjust(bottom=0.13, top=0.85)
 
     # LD plot
-    con0 = ax0.contourf(points_ax1, points_ax2, LD, cmap=colormap, vmin=0, vmax=1, levels=100)
+    n_levels = 100
+    interact_step = 1/n_levels
```
Comment (broncio123) [inline pylds/tools.py:70]: Nice.

File: pylds/tools.py [hunk 5]
```diff
@@ -76,11 +78,11 @@ def ld_plot(LD, LD_gradient, grid_parameters, colormap, interactive, string_titl
...
     vmin = np.nanmin(LD_gradient)
     vmax = np.nanmax(LD_gradient)
-    con1 = ax1.contourf(points_ax1, points_ax2, LD_gradient, cmap=colormap_gradient, levels=100)
+    con1 = ax1.contourf(points_ax1, points_ax2, LD_gradient, cmap=colormap_gradient, levels=n_levels)
     if interactive:
-        @widgets.interact(grad_min=(vmin, vmax-0.01, .01),grad_max=(vmin+0.01, vmax, .01))
+        @widgets.interact(grad_min=(vmin, vmax-interact_step, interact_step),grad_max=(vmin+interact_step, vmax, interact_step))
         def update(grad_min=vmin,grad_max=vmax):
-            grad_max = max(grad_min+0.01, grad_max)
+            grad_max = max(grad_min+interact_step, grad_max)
```
Comment (broncio123) [inline pylds/tools.py]: Thumbs up (emoji)

File: pylds/tools.py (hunk 1 above)
Comment (vkrajnak) [reply pylds/tools.py]: Done

**Comment by vkrajnak** [comment]
Should be all good now

File: LD_Python_Examples.ipynb [hunk 6]
```diff
@@ -1972,6 +1972,13 @@
     "draw_all_lds(LD_forward, LD_backward, grid_parameters, tau, p_value)"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "draw_all_lds returns the handles of the figures and corresponding axes that can be further customised as any other matplotlib plot"
```
Comment (broncio123) [inline LD_Python_Examples.ipynb]: Not clear how to customise a figure using the plot handle.
Add a concrete example or at least a link to a worked example shown somewhere else in "HOW-TOs" section of Notebook.

**Comment by broncio123** [comment]
> Should be all good now

All function changes look appropriate.
Minor improvements in the `LD_Python_Examples` notebook can be done.

Since `draw_all_lds` returns plot handles when running the `LD_Python_Examples` notebook these are outputted shown in the screenshot.
This is easy to fix by directing to a variable, for instance, `obj = draw_all_lds`.

 Also, the `HOW-TOs` section is missing concrete examples on how to use the set of tools in `tools.py` for users to customise their plots. This is important as these sections can be later on used for documentation of the package. 

image

**Review by vkrajnak — COMMENTED** [review]
> > Should be all good now
> 
> All function changes look appropriate.
> Minor improvements in the `LD_Python_Examples` notebook can be done.
> 
> Since `draw_all_lds` returns plot handles when running the `LD_Python_Examples` notebook these are outputted shown in the screenshot.
> This is easy to fix by directing to a variable, for instance, `obj = draw_all_lds`.
> 
> Also, the `HOW-TOs` section is missing concrete examples on how to use the set of tools in `tools.py` for users to customise their plots. This is important as these sections can be later on used for documentation of the package.

The output of `draw_all_lds` is now, as you suggest, directed to a variable.

Regarding the customisation, I'd like to assume basic matplotlib knowledge of the user, same as we do with numpy. Otherwise our examples notebook may turn into some sort of crash course of `contourf`, which is already very well documented by matplotlib. What do you think?

**Comment by vkrajnak** [comment]
> Also, the `HOW-TOs` section is missing concrete examples on how to use the set of tools in `tools.py` for users to customise their plots. This is important as these sections can be later on used for documentation of the package.

Done.

**Pull request merged**